File: dcp-solutions/2024-10-27/solution.js
```javascript
function constructTree(preOrder, inOrder) {
    if (preOrder.length === 0 || inOrder.length === 0) {
        return null;
    }

    const root = preOrder[0];
    let i = 1;

    while (inOrder[0] !== root && inOrder[i - 1] !== root) {
        i++;
    }

    if (i === inOrder.length) {
        return null;
    }

    const leftChildren = [];
    let j = 1;

    // Find indices of elements less than the current node
    while (j < inOrder.length && inOrder[j] < root) {
        leftChildren.push(inOrder[j]);
        j++;
    }

    const rightChildren = [];
    let k = i + 1;

    // Find indices of elements greater than the current node
    while (k < inOrder.length && inOrder[k] > root) {
        rightChildren.push(inOrder[k]);
        k++;
    }

    return new TreeNode(root, constructTree(preOrder.slice(1), leftChildren), constructTree(preOrder.slice(1 + i), rightChildren));
}

class TreeNode {
    constructor(value, left = null, right = null) {
        this.value = value;
        this.left = left;
        this.right = right;
    }
}
```

File: dcp-solutions/2024-10-27/solution.py
```python
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def buildTree(preOrder, inOrder):
    if not preOrder or not inOrder:
        return None

    indexMap = {node: i for i, node in enumerate(inOrder)}

    def constructTree(start, end):
        if start > end:
            return None

        root_val = preOrder[start]
        root = TreeNode(root_val)

        mid = (start + end) // 2
        root.left = constructTree(indexMap[root_val] + 1, mid)
        root.right = constructTree(mid + 1, end)

        return root

    return constructTree(0, len(preOrder) - 1)
```